src/core/lifetime.d

```diff
@@ -2292,3 +2353,186 @@ void _d_delstruct(T)(ref T *p)
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+}
+
+// issue 25552
+@safe unittest
+{
+    int i;
+    struct Nested
+    {
+        int[3] stuff = [0, 1, 2];
+        void foo() { ++i; }
+        ~this() {}
+    }
+
+    static struct NoAssign
+    {
+        int value;
+        @disable void opAssign(typeof(this));
+    }
+
+    static struct NotNested
+    {
+        int before = 42;
+        Nested n;
+        auto after = NoAssign(43);
+    }
+
+    static struct Deep
+    {
+        NotNested nn;
+    }
+
+    static struct Deeper
+    {
+        NotNested[1] nn;
+    }
+
+    static assert(!__traits(isZeroInit, Nested));
+    static assert(!__traits(isZeroInit, NotNested));
+    static assert(!__traits(isZeroInit, Deep));
+    static assert(!__traits(isZeroInit, Deeper));
+
+    auto a = NotNested(1, Nested([3, 4, 5]), NoAssign(2));
+    auto b = move(a);
+    assert(b.n.tupleof[$-1]);
+    assert(a.n.tupleof[$-1] is b.n.tupleof[$-1]);
+    assert(a.n.stuff == a.n.init.stuff);
+    assert(a.before == a.init.before);
+    assert(a.after == a.init.after);
+
+    auto c = Deep(NotNested(1, Nested([3, 4, 5]), NoAssign(2)));
+    auto d = move(c);
+    assert(d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.tupleof[$-1] is d.nn.n.tupleof[$-1]);
+    assert(c.nn.n.stuff == c.nn.n.init.stuff);
+    assert(c.nn.before == c.nn.init.before);
+    assert(c.nn.after == c.nn.init.after);
+
+    auto e = Deeper([NotNested(1, Nested([3, 4, 5]), NoAssign(2))]);
+    auto f = move(e);
+    assert(f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.tupleof[$-1] is f.nn[0].n.tupleof[$-1]);
+    assert(e.nn[0].n.stuff == e.nn[0].n.init.stuff);
+    assert(e.nn[0].before == e.nn[0].init.before);
+    assert(e.nn[0].after == e.nn[0].init.after);
+}
+
+// wipes source after moving (zero init case)
+pragma(inline, true)
+private void wipe(T)(return scope ref T source) @trusted
+{
+    static if (__traits(isStaticArray, T) && hasContextPointers!T)
+    {
+        for (auto i = 0; i < T.length; i++)
+            wipe(source[i]);
+    }
+    else static if (is(T == struct) && hasContextPointers!T)
+    {
+        import core.internal.traits : anySatisfy;
+        static if (anySatisfy!(hasContextPointers, typeof(T.tupleof)))
+        {
+            static foreach (i; 0 .. T.tupleof.length - __traits(isNested, T))
+                wipe(source.tupleof[i]);
+        }
+        else
+        {
+            static if (__traits(isNested, T))
+                enum sz = T.tupleof[$-1].offsetof;
+            else
+                enum sz = T.sizeof;
+            *cast(ubyte[sz]*) &source = 0;
+        }
+    }
+    else
+    {
+        *cast(ubyte[T.sizeof]*) &source = 0;
+    }
+}
+
+// wipes source after moving (initializer case)
+pragma(inline, true)
+private void wipe(T, I)(return scope ref T source, scope ref I initializer) @trusted
+{
+    static if (__traits(isStaticArray, T) && (hasContextPointers!T || !__traits(isPOD, T)))
```

Comment:
I think the POD-ness check isn't required, is it?

Reply:
It is, since the eventual else may do opAssign, and for reasons that are not known to me, presence of custom opAssign does not make a thing not POD :\ Mayhap that needs to be reported?

Reply:
Then moving it to the `hasElaborateAssign` etc. condition would do just fine too, right?

Reply:
Yeah. I think I added that condition some time after making the isPOD test and didn't look at that again.